File: src/cli/builder_cmd.py
"""Typer subcommand group: ``yolo builder {setup,start,stop,status}``.

Manages the on-demand macOS Linux builder VM (see ``builder.py`` and
``docs/handoff-macos-ondemand-builder.md``).  The frictionless replacement
for "run and babysit ``nix run nixpkgs#darwin.linux-builder`` in a terminal":

  * ``setup``  — one-time: wire the Nix daemon to offload aarch64-linux builds
    to the VM (nix.conf builders line, ssh_config, key, launchd service).
  * ``start``  — bring the VM up now (yolo also does this automatically before
    a build that needs it).
  * ``stop``   — bring it down now (a launchd idle-timer also does this).
  * ``status`` — read-only snapshot: set-up? reachable? which conf file?

Registration is via ``app.add_typer(builder_app, name="builder")`` in
cli/__init__.py, mirroring the loopholes/broker groups.
"""

from __future__ import annotations

import typer

from . import builder
from .console import console
from .paths import IS_MACOS


builder_app = typer.Typer(
    help=(
        "Manage the on-demand macOS Linux builder VM.  yolo starts it "
        "automatically before a build that needs it and stops it when idle; "
        "these commands set it up and inspect/override its lifecycle."
    )
)


def _require_macos() -> None:
    if not IS_MACOS:
        console.print(
            "[yellow]The Linux builder is a macOS-only concept.[/yellow]  "
            "On Linux the image builds natively — no builder VM needed."
        )
        raise typer.Exit(0)


@builder_app.command("status")
def builder_status_cmd():
    """Report builder set-up state + reachability (read-only, safe)."""
    _require_macos()
    st = builder.builder_status()

    def mark(ok: bool) -> str:
        return "[green]yes[/green]" if ok else "[red]no[/red]"

    console.print("[bold]macOS Linux builder[/bold]")
    console.print(f"  set up:       {mark(st['done'])}")
    console.print(f"    nix.conf:   {mark(st['nix_builder'])}  ({st['conf_path']})")
    console.print(f"    ssh config: {mark(st['ssh_config'])}")
    console.print(f"    ssh key:    {mark(st['key'])}")
    console.print(
        f"  reachable:    {mark(st['reachable'])}  (port {builder.BUILDER_PORT})"
    )
    console.print()
    if not st["done"]:
        console.print(
            "[yellow]Not set up.[/yellow]  Run [cyan]yolo builder setup[/cyan] "
            "once to wire the Nix daemon to a Linux builder VM."
        )
        raise typer.Exit(1)
    if st["reachable"]:
        console.print("[green]Builder set up and running.[/green]")
        raise typer.Exit(0)
    console.print(
        "[dim]Builder set up but not running — that's normal when idle; "
        "yolo starts it automatically before a build.[/dim]"
    )
    raise typer.Exit(0)


@builder_app.command("start")
def builder_start_cmd():
    """Start the builder VM now (yolo does this automatically before a build)."""
    _require_macos()
    if builder.builder_reachable():
        console.print("[green]Builder already running.[/green]")
        raise typer.Exit(0)
    ok, err = builder.ensure_builder(
        on_progress=lambda m: console.print(f"[dim]{m}[/dim]")
    )
    if ok:
        console.print("[green]Builder is up.[/green]")
        raise typer.Exit(0)
    if err == "not set up":
        console.print(
            "[yellow]Builder not set up.[/yellow]  Run "
            "[cyan]yolo builder setup[/cyan] first."
        )
        raise typer.Exit(1)
    console.print(f"[red]Could not start builder:[/red] {err}")
    raise typer.Exit(1)


@builder_app.command("stop")
def builder_stop_cmd():
    """Stop the builder VM now, reclaiming its RAM (a launchd idle-timer
    also does this automatically)."""
    _require_macos()
    if not builder.builder_reachable():
        console.print("[dim]Builder not running.[/dim]")
        raise typer.Exit(0)
    ok, err = builder.stop_builder()
    if ok:
        console.print("[green]Builder stopped.[/green]")
        raise typer.Exit(0)
    console.print(f"[red]Could not stop builder:[/red] {err}")
    raise typer.Exit(1)


@builder_app.command("setup")
def builder_setup_cmd(
    max_jobs: int = typer.Option(
        4, "--max-jobs", help="Parallel build jobs to allow on the builder."
    ),
    show: bool = typer.Option(
        False, "--show", help="Print the privileged script and exit; don't run it."
    ),
    yes: bool = typer.Option(
        False, "--yes", "-y", help="Skip the confirmation prompt."
    ),
):
    """One-time wiring so the Nix daemon offloads aarch64-linux builds to the VM.

    macOS-only.  Does the work FOR you: explains what will change, shows the
    exact root script, then runs every privileged step in ONE ``sudo`` (a
    single password prompt).  This is an interactive per-run prompt, not a
    sudo-policy change — no NOPASSWD rules, nothing hidden.  ``trusted-users``
    is MERGED (existing entries preserved), never clobbered.
    """
    _require_macos()
    st = builder.builder_setup_state()
    if st["done"]:
        console.print("[green]Builder already set up.[/green]")
        console.print(
            "[dim]Run [cyan]yolo builder status[/cyan] to inspect, or "
            "[cyan]yolo builder start[/cyan] to bring it up.[/dim]"
        )
        raise typer.Exit(0)

    import getpass

    me = getpass.getuser()
    conf = builder._builder_conf_path()
    script = builder.setup_root_script(
        max_jobs, me, builder._current_trusted_users(), conf
    )

    console.print("[bold]Set up the on-demand macOS Linux builder[/bold]\n")
    console.print(
        "macOS can't build the Linux image locally, so Nix offloads to a small "
        "Linux VM.  This wires the Nix daemon to that VM so builds 'just work'; "
        "afterward yolo starts/stops the VM on demand (no terminal to babysit, "
        "no RAM held while idle).\n"
    )
    console.print("[bold]It will, in one sudo:[/bold]")
    console.print(
        f"  • add a [cyan]builders[/cyan] line to [cyan]{conf}[/cyan] (offload aarch64-linux)"
    )
    if builder.trusted_users_line(builder._current_trusted_users(), me) is not None:
        console.print(
            f"  • add [cyan]{me}[/cyan] to [cyan]trusted-users[/cyan] (merged — existing entries kept)"
        )
    else:
        console.print(
            "  • [dim](you're already a trusted user — no trusted-users change)[/dim]"
        )
    console.print(
        f"  • write the ssh host alias [cyan]{builder.SSH_CONFIG_PATH}[/cyan]"
    )
    console.print("  • restart the Nix daemon to apply\n")

    console.print("[dim]The exact root script:[/dim]")
    console.print(f"[dim]{script}[/dim]")

    if show:
        raise typer.Exit(0)

    console.print(
        "[dim]The builder VM itself isn't started here — yolo boots it on "
        "demand ([cyan]nix run nixpkgs#darwin.linux-builder[/cyan]) the first "
        "time a build needs it, which also installs its ssh key (one more sudo "
        "that first time only).[/dim]\n"
    )

    if not yes:
        proceed = typer.confirm("Run the privileged setup now (one sudo prompt)?")
        if not proceed:
            console.print(
                "[dim]Aborted. Re-run when ready, or `--show` to just print it.[/dim]"
            )
            raise typer.Exit(1)

    ok, err = builder.run_setup(max_jobs, me)
    if not ok:
        console.print(f"[red]Setup failed:[/red] {err}")
        raise typer.Exit(1)
    console.print("[green]Builder wired up.[/green]  yolo will start it on demand.")
    raise typer.Exit(0)

File: internal/builder/buildercmd.go
package builder

// The on-demand macOS Linux builder:
// VM lifecycle + reachability, and `yolo builder {setup,start,stop,status}`
// command bodies. macOS can't build the aarch64-linux image locally, so Nix offloads to
// a small Linux VM (nixpkgs#darwin.linux-builder); this brings it up on demand
// and lets a launchd idle-timer stop it.
// The PURE generators (ssh_config block, nix builders line, trusted-users
// merge, the single-sudo root script) live alongside in builder.go
// and are REUSED here, never re-ported. This adds the
// lifecycle orchestration (setup-state probing, ensure/poll/start/stop) and the
// command bodies. Every socket / subprocess / PID-file / platform probe is an
// injectable Deps seam so the orchestration logic is unit-testable on Linux;
// RealDeps wires production. The real VM bring-up stays a Mac-runbook step.

import (
	"fmt"
	"io"
	"regexp"
	"strings"
)

// Proc models a spawned VM process for poll-based liveness
// handle threaded through _poll_until_reachable). Poll returns (returncode, done)
// where done=false means still running (Python's proc.poll() is None).
type Proc interface {
	Poll() (int, bool)
}

// Deps are the injectable primitive seams. RealDeps wires production
// implementations; tests substitute fakes. These mirror the low-level helpers
// in builder.py (socket dial, is_file, nix.conf read, PID file, spawn, killpg)
// so the orchestration (builder_setup_state / ensure_builder / poll) is exact.
type Deps struct {
	// IsMacOS reports sys.platform == "darwin" (paths.IS_MACOS).
	IsMacOS func() bool
	// Reachable reports builder_reachable(): TCP accept on 127.0.0.1:BUILDER_PORT.
	Reachable func() bool
	// FileIsFile reports Path(p).is_file().
	FileIsFile func(path string) bool
	// ReadFileText reads a file best-effort (content, ok). Used by
	// _nix_conf_has_builder + builder_log_tail.
	ReadFileText func(path string) (string, bool)
	// NixCustomConfIncluded reports whether /etc/nix/nix.conf includes
	// nix.custom.conf (storage._nix_custom_conf_included); ok=false when unknown.
	NixCustomConfIncluded func() (bool, bool)
	// CurrentTrustedUsers returns the daemon's effective trusted-users
	// (nix config show), best-effort.
	CurrentTrustedUsers func() []string
	// DetectNixDaemonLabel returns the nix-daemon launchd label (or "", false).
	DetectNixDaemonLabel func() (string, bool)
	// HostUser is getpass.getuser().
	HostUser func() string
	// RunSetupScript pipes the root script to `sudo bash -s` (tty inherited for
	// the password prompt). Returns (returncode, ok) — ok=false on spawn error.
	RunSetupScript func(script string) (int, bool)
	// StartVMForeground runs `nix run nixpkgs#darwin.linux-builder` inheriting
	// the TTY (first_boot_interactive); returns an error (KeyboardInterrupt maps
	// to nil in the caller) or nil.
	StartVMForeground func() error
	// StartVMDetached spawns the VM detached and records the PID file; returns
	// (Proc, err). A nil Proc with nil err means "already running".
	StartVMDetached func() (Proc, error)
	// ReadBuilderPID reads the recorded VM PID (pid, ok).
	ReadBuilderPID func() (int, bool)
	// PIDIsLive reports os.kill(pid, 0) success.
	PIDIsLive func(pid int) bool
	// StopVM terminates the detached VM (killpg → kill fallback) and removes the
	// PID file. Returns (ok, errMsg).
	StopVM func() (bool, string)
	// Sleep/Now drive _poll_until_reachable (injected so tests avoid real waits).
	Sleep func(seconds float64)
	Now   func() float64
	// Confirm prompts (typer.confirm) for the setup step.
	Confirm func(prompt string) bool
	// Out receives the human output (rich markup stripped, parity on text).
	Out io.Writer
}

// Poll timing.
const (
	builderStartTimeoutS = 90.0
	builderPollIntervalS = 1.0
)

var richTagRe = regexp.MustCompile(`\[/?[a-zA-Z][^\]]*\]`)

type printer struct{ w io.Writer }

func (p printer) print(msg string)          { fmt.Fprintln(p.w, richTagRe.ReplaceAllString(msg, "")) }
func (p printer) printf(f string, a ...any) { p.print(fmt.Sprintf(f, a...)) }

type SetupState struct {
	SSHConfig  bool
	NixBuilder bool
	Key        bool
	Done       bool
}

type Status struct {
	SetupState
	Reachable   bool
	ConfPath    string
	DaemonLabel string // "" when none
}

// nix.conf.
func confPath(deps Deps) string {
	if inc, ok := deps.NixCustomConfIncluded(); ok && inc {
		return "/etc/nix/nix.custom.conf"
	}
	return "/etc/nix/nix.conf"
}

// uncommented `builders` line naming both aarch64-linux and the ssh host alias.
func nixConfHasBuilder(deps Deps) bool {
	conf := confPath(deps)
	if !deps.FileIsFile(conf) {
		return false
	}
	text, ok := deps.ReadFileText(conf)
	if !ok {
		return false
	}
	for _, raw := range strings.Split(text, "\n") {
		s := strings.TrimSpace(raw)
		if strings.HasPrefix(s, "#") || !strings.HasPrefix(s, "builders") {
			continue
		}
		if strings.Contains(s, "aarch64-linux") && strings.Contains(s, BuilderSSHHost) {
			return true
		}
	}
	return false
}

// BuilderSetupState probes setup state without touching the VM. Mirrors
// builder_setup_state.
func BuilderSetupState(deps Deps) SetupState {
	sshOK := deps.FileIsFile(SSHConfigPath())
	keyOK := deps.FileIsFile(BuilderKeyPath)
	nixOK := nixConfHasBuilder(deps)
	return SetupState{
		SSHConfig:  sshOK,
		NixBuilder: nixOK,
		Key:        keyOK,
		Done:       sshOK && nixOK,
	}
}

// BuilderStatus is the full read-only snapshot.
func BuilderStatus(deps Deps) Status {
	label, _ := deps.DetectNixDaemonLabel()
	return Status{
		SetupState:  BuilderSetupState(deps),
		Reachable:   deps.Reachable(),
		ConfPath:    confPath(deps),
		DaemonLabel: label,
	}
}

// RunSetup does the whole privileged wiring in ONE sudo.
// builds the root script (reusing internal/builder + the resolved daemon label)
// and pipes it to `sudo bash -s`. Returns (ok, errMsg).
func RunSetup(deps Deps, maxJobs int, me string) (bool, string) {
	label, _ := deps.DetectNixDaemonLabel()
	script := SetupRootScript(maxJobs, me, deps.CurrentTrustedUsers(), confPath(deps), label)
	rc, ok := deps.RunSetupScript(script)
	if !ok {
		return false, "privileged setup failed to run"
	}
	if rc != 0 {
		return false, fmt.Sprintf("privileged setup exited %d", rc)
	}
	return true, ""
}

// FirstBootInteractive runs the VM's one-time first boot in the foreground.
// installed (or reachability). Returns (ok, errMsg).
func FirstBootInteractive(deps Deps) (bool, string) {
	if err := deps.StartVMForeground(); err != nil {
		return false, err.Error()
	}
	if deps.FileIsFile(BuilderKeyPath) || deps.Reachable() {
		return true, ""
	}
	return false, "ssh key still not installed after first boot"
}

// EnsureBuilder makes the builder ready to accept a build, starting it if
// needed.
// onProgress is optional (nil = silent).
func EnsureBuilder(deps Deps, onProgress func(string)) (bool, string) {
	if !deps.IsMacOS() {
		return false, "not macOS"
	}
	if deps.Reachable() {
		return true, ""
	}
	state := BuilderSetupState(deps)
	if !state.Done {
		return false, "not set up"
	}
	if !state.Key {
		return false, "needs first-boot"
	}
	if onProgress != nil {
		onProgress("starting Linux builder VM (first boot downloads it)…")
	}
	proc, err := deps.StartVMDetached()
	if err != nil {
		return false, "could not start builder: " + err.Error()
	}
	ok, reason := pollUntilReachable(deps, proc, onProgress)
	if ok {
		return true, ""
	}
	return false, reason
}

// pollUntilReachable polls the builder's SSH port until it answers, the child
// dies, or timeout.
// short-circuit with a log-tail-derived reason, and the final re-check).
func pollUntilReachable(deps Deps, proc Proc, onProgress func(string)) (bool, string) {
	deadline := deps.Now() + builderStartTimeoutS
	waited := 0
	for deps.Now() < deadline {
		if deps.Reachable() {
			return true, ""
		}
		if proc != nil {
			if _, done := proc.Poll(); done {
				last := lastLogLine(deps)
				if last == "" {
					last = "no output"
				}
				return false, "builder process exited early (" + last + ")"
			}
		}
		if onProgress != nil {
			onProgress(fmt.Sprintf("waiting for Linux builder VM to boot… (%ds)", waited))
		}
		deps.Sleep(builderPollIntervalS)
		// Python: waited += int(interval_s) or 1
		inc := int(builderPollIntervalS)
		if inc == 0 {
			inc = 1
		}
		waited += inc
	}
	if deps.Reachable() {
		return true, ""
	}
	hint := ""
	if last := lastLogLine(deps); last != "" {
		hint = "; last log line: " + last
	}
	return false, fmt.Sprintf("not reachable within %ds%s", int(builderStartTimeoutS), hint)
}

// lastLogLine returns the last non-empty line of the builder log tail (matching
// builder_log_tail(...).splitlines()[-1] usage). "" when the log is empty.
func lastLogLine(deps Deps) string {
	text, ok := deps.ReadFileText(BuilderLogFilePath())
	if !ok {
		return ""
	}
	// builder_log_tail: last 12 lines joined + strip; callers take [-1]. The net
	// effect for the reason string is the last non-empty line of the file.
	lines := strings.Split(text, "\n")
	tail := lines
	if len(tail) > 12 {
		tail = tail[len(tail)-12:]
	}
	joined := strings.TrimSpace(strings.Join(tail, "\n"))
	if joined == "" {
		return ""
	}
	parts := strings.Split(joined, "\n")
	return parts[len(parts)-1]
}
